File: client/src/heightmap.ts
import * as THREE from 'three';
import { publicAssetPath } from './publicAssets';
import {
  HEIGHTMAP_BIN_PUBLIC_PATH,
  HEIGHTMAP_MAX_X,
  HEIGHTMAP_MAX_Z,
  HEIGHTMAP_MIN_X,
  HEIGHTMAP_MIN_Z,
  HEIGHTMAP_SIZE,
  TERRAIN_MAX_WALKABLE_SLOPE_DEGREES,
} from './heightmapMeta';

export {
  HEIGHTMAP_MAX_X,
  HEIGHTMAP_MAX_Z,
  HEIGHTMAP_MIN_X,
  HEIGHTMAP_MIN_Z,
  HEIGHTMAP_SIZE,
  TERRAIN_MAX_WALKABLE_SLOPE_DEGREES,
};

const TERRAIN_MAX_WALKABLE_SLOPE = Math.tan(
  THREE.MathUtils.degToRad(TERRAIN_MAX_WALKABLE_SLOPE_DEGREES),
);

let heights: Float32Array | null = null;
let walkable: Uint8Array | null = null;
let loadPromise: Promise<void> | null = null;

function requireHeights(): Float32Array {
  if (!heights) {
    throw new Error('Heightmap not loaded — call await initHeightmap() before sampling');
  }
  return heights;
}

function requireWalkable(): Uint8Array {
  if (!walkable) {
    throw new Error('Heightmap not loaded — call await initHeightmap() before sampling');
  }
  return walkable;
}

/** Decode HM01 binary (see scripts/convert-heightmap-binary.mjs). */
export function loadHeightmapFromArrayBuffer(buffer: ArrayBuffer): void {
  const view = new DataView(buffer);
  const magic = String.fromCharCode(
    view.getUint8(0),
    view.getUint8(1),
    view.getUint8(2),
    view.getUint8(3),
  );
  if (magic !== 'HM01') {
    throw new Error(`Bad heightmap magic: ${magic}`);
  }
  const size = view.getUint32(4, true);
  if (size !== HEIGHTMAP_SIZE) {
    throw new Error(`Heightmap size ${size} != expected ${HEIGHTMAP_SIZE}`);
  }

  const count = size * size;
  const heightsOffset = 4 + 4 + 4 * 6;
  const nextHeights = new Float32Array(count);
  for (let i = 0; i < count; i += 1) {
    nextHeights[i] = view.getFloat32(heightsOffset + i * 4, true);
  }

  const walkOffset = heightsOffset + count * 4;
  const walkBytes = Math.ceil(count / 8);
  const nextWalkable = new Uint8Array(buffer, walkOffset, walkBytes);

  heights = nextHeights;
  walkable = new Uint8Array(nextWalkable); // copy out of the shared buffer
}

/**
 * Load collision heightmap from the public binary asset.
 * Safe to call multiple times; concurrent callers share one fetch.
 */
export function initHeightmap(
  url = publicAssetPath(HEIGHTMAP_BIN_PUBLIC_PATH),
): Promise<void> {
  if (heights && walkable) return Promise.resolve();
  if (!loadPromise) {
    loadPromise = fetch(url)
      .then(response => {
        if (!response.ok) {
          throw new Error(`Failed to fetch heightmap: ${response.status} ${url}`);
        }
        return response.arrayBuffer();
      })
      .then(buffer => {
        loadHeightmapFromArrayBuffer(buffer);
      })
      .catch(error => {
        loadPromise = null;
        throw error;
      });
  }
  return loadPromise;
}

/** Test/helper: inject samples without fetch. */
export function loadHeightmapFromBytes(bytes: Uint8Array): void {
  const copy = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(copy).set(bytes);
  loadHeightmapFromArrayBuffer(copy);
}

export function isHeightmapReady(): boolean {
  return heights != null && walkable != null;
}

export function terrainHeightAt(position: THREE.Vector3): number {
  return sampleHeight(position.x, position.z);
}

export function sampleHeight(x: number, z: number): number {
  const u = THREE.MathUtils.clamp((x - HEIGHTMAP_MIN_X) / (HEIGHTMAP_MAX_X - HEIGHTMAP_MIN_X), 0, 1);
  const v = THREE.MathUtils.clamp((z - HEIGHTMAP_MIN_Z) / (HEIGHTMAP_MAX_Z - HEIGHTMAP_MIN_Z), 0, 1);
  const gx = u * (HEIGHTMAP_SIZE - 1);
  const gz = v * (HEIGHTMAP_SIZE - 1);
  const x0 = Math.floor(gx);
  const z0 = Math.floor(gz);
  const x1 = Math.min(x0 + 1, HEIGHTMAP_SIZE - 1);
  const z1 = Math.min(z0 + 1, HEIGHTMAP_SIZE - 1);
  const tx = gx - x0;
  const tz = gz - z0;
  const h00 = heightAtIndex(x0, z0);
  const h10 = heightAtIndex(x1, z0);
  const h01 = heightAtIndex(x0, z1);
  const h11 = heightAtIndex(x1, z1);
  const h0 = THREE.MathUtils.lerp(h00, h10, tx);
  const h1 = THREE.MathUtils.lerp(h01, h11, tx);
  return THREE.MathUtils.lerp(h0, h1, tz);
}

export function isTerrainWalkableAt(x: number, z: number): boolean {
  const u = THREE.MathUtils.clamp((x - HEIGHTMAP_MIN_X) / (HEIGHTMAP_MAX_X - HEIGHTMAP_MIN_X), 0, 1);
  const v = THREE.MathUtils.clamp((z - HEIGHTMAP_MIN_Z) / (HEIGHTMAP_MAX_Z - HEIGHTMAP_MIN_Z), 0, 1);
  const gridX = Math.round(u * (HEIGHTMAP_SIZE - 1));
  const gridZ = Math.round(v * (HEIGHTMAP_SIZE - 1));
  return isWalkableIndex(gridZ * HEIGHTMAP_SIZE + gridX)
    && terrainSlopeAt(x, z) <= TERRAIN_MAX_WALKABLE_SLOPE;
}

export function terrainSlopeAt(x: number, z: number): number {
  const cellX = (HEIGHTMAP_MAX_X - HEIGHTMAP_MIN_X) / (HEIGHTMAP_SIZE - 1);
  const cellZ = (HEIGHTMAP_MAX_Z - HEIGHTMAP_MIN_Z) / (HEIGHTMAP_SIZE - 1);
  const left = sampleHeight(x - cellX, z);
  const right = sampleHeight(x + cellX, z);
  const down = sampleHeight(x, z - cellZ);
  const up = sampleHeight(x, z + cellZ);
  const dhdx = (right - left) / (cellX * 2);
  const dhdz = (up - down) / (cellZ * 2);
  return Math.hypot(dhdx, dhdz);
}

function heightAtIndex(x: number, z: number): number {
  return requireHeights()[z * HEIGHTMAP_SIZE + x];
}

function isWalkableIndex(index: number): boolean {
  const byte = requireWalkable()[index >> 3];
  return (byte & (0x80 >> (index & 7))) !== 0;
}

File: server/spacetimedb/src/heightmap.rs
use crate::common::Vector3;
use std::sync::OnceLock;

/// Expected grid size written by `scripts/bake-terrain-collision.mjs`.
pub const HEIGHTMAP_SIZE: usize = 513;
pub const TERRAIN_MAX_WALKABLE_SLOPE_DEGREES: f32 = 70.0;

const TERRAIN_MAX_WALKABLE_SLOPE: f32 = 2.74747742; // tan(70 degrees)

/// Baked collision surface (HM01). Produced by `scripts/bake-terrain-collision.mjs`.
/// World bounds are read from the binary header so re-bakes cannot desync constants.
static HEIGHTMAP_BIN: &[u8] = include_bytes!("heightmap.bin");

struct HeightmapData {
    min_x: f32,
    max_x: f32,
    min_z: f32,
    max_z: f32,
    heights: Vec<f32>,
    walkable: Vec<u8>,
}

static HEIGHTMAP: OnceLock<HeightmapData> = OnceLock::new();

fn heightmap() -> &'static HeightmapData {
    HEIGHTMAP.get_or_init(|| parse_heightmap_bin(HEIGHTMAP_BIN).expect("valid embedded heightmap.bin"))
}

fn parse_heightmap_bin(bytes: &[u8]) -> Result<HeightmapData, String> {
    if bytes.len() < 32 {
        return Err("heightmap.bin too short".into());
    }
    if &bytes[0..4] != b"HM01" {
        return Err("bad heightmap magic".into());
    }
    let size = u32::from_le_bytes(bytes[4..8].try_into().unwrap()) as usize;
    if size != HEIGHTMAP_SIZE {
        return Err(format!("heightmap size {size} != {HEIGHTMAP_SIZE}"));
    }

    let min_x = f32::from_le_bytes(bytes[8..12].try_into().unwrap());
    let max_x = f32::from_le_bytes(bytes[12..16].try_into().unwrap());
    let min_z = f32::from_le_bytes(bytes[16..20].try_into().unwrap());
    let max_z = f32::from_le_bytes(bytes[20..24].try_into().unwrap());
    // bytes 24..32 are minH/maxH — unused for sampling UV

    let count = size * size;
    let heights_offset = 8 + 24;
    let walk_offset = heights_offset + count * 4;
    let walk_bytes = count.div_ceil(8);
    if bytes.len() < walk_offset + walk_bytes {
        return Err("heightmap.bin truncated".into());
    }

    let mut heights = Vec::with_capacity(count);
    for i in 0..count {
        let start = heights_offset + i * 4;
        heights.push(f32::from_le_bytes(bytes[start..start + 4].try_into().unwrap()));
    }

    Ok(HeightmapData {
        min_x,
        max_x,
        min_z,
        max_z,
        heights,
        walkable: bytes[walk_offset..walk_offset + walk_bytes].to_vec(),
    })
}

pub fn terrain_height_at(position: &Vector3) -> f32 {
    sample_height(position.x, position.z)
}

pub fn sample_height(x: f32, z: f32) -> f32 {
    let hm = heightmap();
    let u = ((x - hm.min_x) / (hm.max_x - hm.min_x)).clamp(0.0, 1.0);
    let v = ((z - hm.min_z) / (hm.max_z - hm.min_z)).clamp(0.0, 1.0);
    let gx = u * (HEIGHTMAP_SIZE as f32 - 1.0);
    let gz = v * (HEIGHTMAP_SIZE as f32 - 1.0);
    let x0 = gx.floor() as usize;
    let z0 = gz.floor() as usize;
    let x1 = (x0 + 1).min(HEIGHTMAP_SIZE - 1);
    let z1 = (z0 + 1).min(HEIGHTMAP_SIZE - 1);
    let tx = gx - x0 as f32;
    let tz = gz - z0 as f32;
    let h00 = height_at_index(x0, z0);
    let h10 = height_at_index(x1, z0);
    let h01 = height_at_index(x0, z1);
    let h11 = height_at_index(x1, z1);
    let h0 = h00 + (h10 - h00) * tx;
    let h1 = h01 + (h11 - h01) * tx;
    h0 + (h1 - h0) * tz
}

pub fn is_terrain_walkable_at(x: f32, z: f32) -> bool {
    let hm = heightmap();
    let u = ((x - hm.min_x) / (hm.max_x - hm.min_x)).clamp(0.0, 1.0);
    let v = ((z - hm.min_z) / (hm.max_z - hm.min_z)).clamp(0.0, 1.0);
    let grid_x = (u * (HEIGHTMAP_SIZE as f32 - 1.0)).round() as usize;
    let grid_z = (v * (HEIGHTMAP_SIZE as f32 - 1.0)).round() as usize;
    let index = grid_z * HEIGHTMAP_SIZE + grid_x;
    is_walkable_index(index) && terrain_slope_at(x, z) <= TERRAIN_MAX_WALKABLE_SLOPE
}

pub fn terrain_slope_at(x: f32, z: f32) -> f32 {
    let hm = heightmap();
    let cell_x = (hm.max_x - hm.min_x) / (HEIGHTMAP_SIZE as f32 - 1.0);
    let cell_z = (hm.max_z - hm.min_z) / (HEIGHTMAP_SIZE as f32 - 1.0);
    let left = sample_height(x - cell_x, z);
    let right = sample_height(x + cell_x, z);
    let down = sample_height(x, z - cell_z);
    let up = sample_height(x, z + cell_z);
    let dhdx = (right - left) / (cell_x * 2.0);
    let dhdz = (up - down) / (cell_z * 2.0);
    (dhdx * dhdx + dhdz * dhdz).sqrt()
}

fn height_at_index(x: usize, z: usize) -> f32 {
    heightmap().heights[z * HEIGHTMAP_SIZE + x]
}

fn is_walkable_index(index: usize) -> bool {
    let byte = heightmap().walkable[index >> 3];
    (byte & (0x80 >> (index & 7))) != 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn samples_inside_generated_range() {
        let h = sample_height(0.0, 0.0);
        assert!(h >= 0.0);
    }

    #[test]
    fn embedded_bin_parses_bounds_from_header() {
        let data = parse_heightmap_bin(HEIGHTMAP_BIN).expect("parse");
        assert_eq!(data.heights.len(), HEIGHTMAP_SIZE * HEIGHTMAP_SIZE);
        assert!(data.max_x > data.min_x);
        assert!(data.max_z > data.min_z);
        // Matches client heightmapMeta for the current bake (approx; f32 binary).
        assert!((data.min_x - (-1574.03)).abs() < 0.05);
        assert!((data.min_z - (-1218.91)).abs() < 0.05);
    }
}
